Userland/Utilities/crypto.cpp

```diff
@@ -0,0 +1,239 @@
+        fill_with_random(buffer);
+
+        Timings timing_result { .unit_bytes = size };
+        warn("Running benchmark for {} with size {} for ~{}ms...", name, size, TimeSlicePerSize.to_milliseconds());
+        auto total_timer = Core::ElapsedTimer::start_new(Core::TimerType::Precise);
+        for (; total_timer.elapsed_time() < TimeSlicePerSize;) {
+            auto timer = Core::ElapsedTimer::start_new(Core::TimerType::Precise);
+            func(buffer);
+            auto elapsed = timer.elapsed_time();
+            timing_result.max_us = max(timing_result.max_us, elapsed.to_microseconds());
+            timing_result.min_us = min(timing_result.min_us, elapsed.to_microseconds());
+            timing_result.total_us += elapsed.to_microseconds();
+            timing_result.count++;
+        }
+        g_all_timings.ensure(name).set(size, timing_result);
+        warnln("{}ms, {} ops, {}/s", total_timer.elapsed_milliseconds(), timing_result.count, human_readable_quantity(timing_result.unit_bytes * timing_result.count / TimeSlicePerSize.to_microseconds() * 1'000'000));
+    }
+}
+
+template<typename Algorithm>
+static ErrorOr<void> run_hash_benchmark(StringView name)
+{
+    run_benchmark_with_all_sizes(name, [](auto& buffer) {
+        auto digest = Algorithm::hash(buffer);
+        use(digest);
+    });
+    return {};
+}
+
+template<typename Algorithm>
+static ErrorOr<void> run_checksum_benchmark(StringView name)
+{
+    run_benchmark_with_all_sizes(name, [](auto& buffer) {
+        Algorithm checksum;
+        checksum.update(buffer);
+        auto digest = checksum.digest();
+        use(digest);
+    });
+    return {};
+}
+
+template<typename Algorithm>
+static ErrorOr<void> run_auth_benchmark(StringView name)
+{
+    auto key = TRY(ByteBuffer::create_uninitialized(128));
+    fill_with_random(key);
+    run_benchmark_with_all_sizes(name, [&](auto& buffer) {
+        Algorithm auth(key.bytes());
+        if constexpr (requires { auth.process(buffer.bytes()); }) {
+            auto tag = auth.process(buffer.bytes());
+            use(tag);
+        } else if constexpr (requires { auth.process(buffer.bytes(), buffer.bytes()); }) {
+            auto tag = auth.process(buffer.bytes(), buffer.bytes());
+            use(tag);
+        } else {
+            auth.update(buffer.bytes());
+            auto digest = auth.digest();
+            use(digest);
+        }
+    });
+    return {};
+}
+
+template<typename Algorithm, typename... Nonce>
+static ErrorOr<void> run_cipher_benchmark(StringView name, size_t key_bits, Nonce...)
+{
+    constexpr auto needs_nonce = sizeof...(Nonce) == 1;
+    auto key = TRY(ByteBuffer::create_uninitialized(key_bits / 8));
+    fill_with_random(key);
+
+    auto out_buffer = TRY(ByteBuffer::create_uninitialized(16 * MiB));
+
+    auto iv = TRY(ByteBuffer::create_uninitialized(key_bits / 8));
+    fill_with_random(iv);
+
+    auto nonce = TRY(ByteBuffer::create_uninitialized(12));
```

Comment:
nit: `12` looks like an unexplained magic number to this uniformed observer

Reply:
it really is a magic number, it's the nonce size required by the only cipher that needs one.
I'll add a comment.